dependency-specification.rst

```diff
@@ -0,0 +1,324 @@
+
+Concepts
+--------
+
+A dependency specification always specifies a distribution name. It may
+include extras, which expand the dependencies of the named distribution to
+enable optional features. The version installed can be controlled using
+version limits, or giving the URL to a specific artifact to install. Finally
+the dependency can be made conditional using environment markers.
+
+Grammar
+-------
+
+We first cover the grammar briefly and then drill into the semantics of each
+section later.
+
+A distribution specification is written in ASCII text. We use ABNF [#abnf]_ to
+provide a precise grammar. The grammar covers just the specification itself.
+The expectation is that the specification will be embedded into a larger
+system which offers framing such as comments, multiple line support via
+continuations, or other such features.
+
+Versions may be specified according to the PEP-440 [#pep440]_ rules. (Note:
+URI is defined in std-66 [#std66]_::
+
+    version-cmp   = "<" / "<=" / "!=" / "==" / ">=" / ">"
+    version       = 1*( DIGIT / ALPHA / "-" / "_" / "." / "*" )
+    version-inner = version-cmp version *(',' version-cmp version)
+    versionspec   = ("(" version-inner ")") / version-inner
+    urlspec       = "@" URI
+
+Environment markers allow making a specification only take effect in some
+environments::
+
+    marker-op     = version-cmp / "in" / "not in"
+    python-str-c  = WSP / ALPHA / DIGIT / "(" / ")" / "." / "{" / "}" /
+                    "-" / "_" / "*"
+    python-str    = "'" *(python-str-c / DQUOTE) "'"
+    python-str    =/ DQUOTE *(python-str-c / "'") DQUOTE
+    marker-var    = python-str / "python_version" / "python_full_version" /
+                    "os_name"" / "sys_platform" / "platform_release" /
+                    "platform_version" / "platform_machine" /
+                    "platform_python_implementation" / "implementation_name" /
+                    "implementation_version" / "platform_dist_name"
+                    "platform_dist_version" / "platform_dist_id"
+    marker-var    =/ "extra" ; ONLY when defined by a containing layer
+    marker-expr   = "(" *WSP marker *WSP ")"
+                  =/ (marker-var [*WSP marker-op *WSP marker-var])
+    marker        = *WSP marker-expr *( *WSP ("and" / "or") *WSP marker-expr)
+    name-marker   = ";" *WSP marker
+    url-marker    = ";" 1*WSP marker
+
+Optional components of a distribution may be specified using the extras
+field::
+
+    identifier    = 1*( DIGIT / ALPHA / "-" / "_" )
+    name          = identifier
+    extras        = "[" identifier *("," identifier) "]"
+
+Giving us a rule for name based requirements::
+
+    name_req      = name [*WSP extras] [*WSP versionspec] [*WSP name-marker]
+
+And a rule for direct reference specifications::
+
+    url_req       = name [*WSP extras] urlspec [*WSP url-marker]
+
+Leading to the unified rule that can specify a dependency::
+
+    specification = name_req / location_req
```

Comment:
This should probably support leading and trailing whitespace which is just ignored as well yea?